### PR TITLE
Add `KnownLayout`, remove `Sized` from some APIs

### PR DESCRIPTION
**Pull request by joshlf**
Makes progress on #29

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1936,22 +1941,30 @@ pub unsafe trait FromBytes: FromZeros {
...
+    /// let packet = Packet::ref_from_prefix(bytes).unwrap();
+    ///
+    /// assert_eq!(packet.header.src_port, [0, 1]);
+    /// assert_eq!(packet.header.dst_port, [2, 3]);
+    /// assert_eq!(packet.header.length, [4, 5]);
+    /// assert_eq!(packet.header.checksum, [6, 7]);
+    /// assert_eq!(packet.body, [[8, 9], [10, 11], [12, 13]]);
     /// ```
     #[inline]
     fn ref_from_prefix(bytes: &[u8]) -> Option<&Self>
```
Comment (joshlf) [inline src/lib.rs:1963]: Need to figure out how to reword this paragraph from the doc comment:

```rust
    /// `ref_from_prefix` returns a reference to the first `size_of::<Self>()`
    /// bytes of `bytes`. If `bytes.len() < size_of::<Self>()` or `bytes` is not
    /// aligned to `align_of::<Self>()`, this returns `None`.
```

File: src/lib.rs [hunk 2]
```diff
@@ -1985,9 +1998,9 @@ pub unsafe trait FromBytes: FromZeros {
     #[inline]
     fn ref_from_suffix(bytes: &[u8]) -> Option<&Self>
```
Comment (joshlf) [inline src/lib.rs:1999]: Need to figure out how to reword this paragraph from the doc comment:

```rust
    /// `ref_from_suffix` returns a reference to the last `size_of::<Self>()`
    /// bytes of `bytes`. If `bytes.len() < size_of::<Self>()` or the suffix of
    /// `bytes` is not aligned to `align_of::<Self>()`, this returns `None`.
```

File: src/lib.rs (hunk 1 above)
Comment (joshlf) [reply src/lib.rs:1963]: Added a TODO in #29 to block release on updating this doc comment, so we don't have to resolve this now.

File: src/lib.rs (hunk 2 above)
Comment (joshlf) [reply src/lib.rs:1999]: Added a TODO in #29 to block release on updating this doc comment, so we don't have to resolve this now.

**Review by jswrenn — APPROVED** [review]

**Pull request merged**